include/libtorrent/packet_pool.hpp

```diff
@@ -0,0 +1,180 @@
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_PACKET_POOL_HPP
+#define TORRENT_PACKET_POOL_HPP
+
+#include "libtorrent/config.hpp"
+
+#include "libtorrent/aux_/vector.hpp"
+
+namespace libtorrent
+{
+
+	// used for out-of-order incoming packets
+	// as well as sent packets that are waiting to be ACKed
+	struct packet
+	{
+		// the last time this packet was sent
+		time_point send_time;
+
+		// the number of bytes actually allocated in 'buf'
+		std::uint16_t allocated;
+
+		// the size of the buffer 'buf' points to
+		std::uint16_t size;
+
+		// this is the offset to the payload inside the buffer
+		// this is also used as a cursor to describe where the
+		// next payload that hasn't been consumed yet starts
+		std::uint16_t header_size;
+
+		// the number of times this packet has been sent
+		std::uint8_t num_transmissions:6;
+
+		// true if we need to send this packet again. All
+		// outstanding packets are marked as needing to be
+		// resent on timeouts
+		bool need_resend:1;
+
+		// this is set to true for packets that were
+		// sent with the DF bit set (Don't Fragment)
+		bool mtu_probe:1;
+
+	#if TORRENT_USE_ASSERTS
+		int num_fast_resend;
+	#endif
+
+		// the actual packet buffer
+		std::uint8_t buf[1];
+	};
+
+
+	struct packet_deleter
+	{
+		void operator()(packet* handle) const
+		{
+			handle->~packet();
+			std::free(handle);
+		}
+	};
+
+	using packet_ptr = std::unique_ptr<packet, packet_deleter>;
+
+	template<int ALLOCATE_SIZE>
+	struct TORRENT_EXTRA_EXPORT packet_slab
+	{
+		static const int allocate_size{ ALLOCATE_SIZE };
+
+		explicit packet_slab(std::size_t limit) : m_limit(limit) { m_storage.reserve(m_limit); }
+
+		packet_slab(const packet_slab&) = delete;
+
+		void try_push_back(packet_ptr &p)
+		{
+			if (m_storage.size() < m_limit)
+				m_storage.push_back(std::move(p));
+		}
+
+		packet_ptr alloc()
+		{
+			if (m_storage.empty())
+				return packet_ptr(static_cast<packet*>(std::malloc(sizeof(packet) + allocate_size)));
+			else
+			{
+				auto ret = std::move(m_storage.back());
+				m_storage.pop_back();
+				return ret;
+			}
+		}
+
+		void flush() { m_storage.resize(0); }
+	private:
+		const std::size_t m_limit;
+		aux::vector<packet_ptr, std::size_t> m_storage;
```

Comment:
i thought all std::vector inside libtorrent should migrate to these implementation (it has some usefull asserts?)